plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/util/HudiAvroSerializer.java

```
@@ -0,0 +1,290 @@
+
+public class HudiAvroSerializer {
+    private static final int[] NANO_FACTOR = {
+            -1, // 0, no need to multiply
+            100_000_000, // 1 digit after the dot
+            10_000_000, // 2 digits after the dot
+            1_000_000, // 3 digits after the dot
+            100_000, // 4 digits after the dot
+            10_000, // 5 digits after the dot
+            1000, // 6 digits after the dot
+            100, // 7 digits after the dot
+            10, // 8 digits after the dot
+            1, // 9 digits after the dot
+    };
+
+    private static final AvroDecimalConverter DECIMAL_CONVERTER = new AvroDecimalConverter();
+
+    private final List<HiveColumnHandle> columnHandles;
+    private final List<Type> columnTypes;
+    private final Schema schema;
+
+    public HudiAvroSerializer(List<HiveColumnHandle> columnHandles) {
+        this.columnHandles = columnHandles;
+        this.columnTypes = columnHandles.stream().map(HiveColumnHandle::getType).toList();
+        this.schema = constructSchema(columnHandles.stream().map(HiveColumnHandle::getName).toList(),
+                columnHandles.stream().map(HiveColumnHandle::getHiveType).toList(), false);
+    }
+
+    public IndexedRecord serialize(Page page, int position) {
+        IndexedRecord record = new GenericData.Record(schema);
+        for (int i = 0; i < columnTypes.size(); i++) {
+            Object value = getValue(page, i, position);
+            record.put(i, value);
+        }
+        return record;
+    }
+
+    public Object getValue(Page page, int channel, int position) {
+        return columnTypes.get(channel).getObjectValue(null, page.getBlock(channel), position);
+    }
+
+    public void buildRecordInPage(PageBuilder pageBuilder, IndexedRecord record,
+                                  Map<Integer, String> partitionValueMap, boolean SkipMetaColumns) {
+        pageBuilder.declarePosition();
+        int startChannel = SkipMetaColumns ? HOODIE_META_COLUMNS.size() : 0;
```

This causes a bug when there are hudi meta columns present in the table schema itself and are selected as the pagebuilder blocks will contain entries for these columns, however they will always be skipped which can cause index out of bounds errors and/or type mismatch errors that get swallowed due to the block entries and type entries being out of sync. The proposed solution is to instead of passing in a boolean of whether or not to skipMetaColumns pass in an int from HudiSnapshotPageSource caller which is the appropriate number of meta columns to be skipped, which is just the total number of meta columns that exist - number of meta columns in the data columns.